drivers/adc/adc_cc32xx.c

```diff
@@ -0,0 +1,291 @@
+	return 0;
+}
+
+static int adc_cc32xx_channel_setup(const struct device *dev,
+				    const struct adc_channel_cfg *channel_cfg)
+{
+	struct adc_cc32xx_data *data = dev->data;
+	const struct adc_cc32xx_cfg *config = dev->config;
+	uint8_t ch = channel_cfg->channel_id;
+
+	if (ch >= ARRAY_SIZE(s_chPin)) {
+		LOG_ERR("Channel is not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Acquisition time is not valid");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Gain is not valid");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Reference is not valid");
+		return -EINVAL;
+	}
+
+	data->ch = ch;
+
+	LOG_DBG("Setup %d", ch);
+
+	MAP_ADCChannelDisable(config->base, s_channel[ch]);
+	MAP_ADCIntDisable(config->base, s_channel[ch], ISR_MASK);
+	MAP_PinDirModeSet(s_chPin[ch], PIN_DIR_MODE_IN);
+	MAP_PinTypeADC(s_chPin[ch], PIN_MODE_255);
+
+	return 0;
+}
+
+static int cc32xx_read(const struct device *dev,
+		       const struct adc_sequence *sequence,
+		       bool asynchronous,
+		       struct k_poll_signal *sig)
+{
+	struct adc_cc32xx_data *data = dev->data;
+	uint32_t channels = sequence->channels;
+	uint32_t index = find_lsb_set(channels) - 1;
+	int rv;
+
+	if (sequence->resolution != 12) {
+		LOG_ERR("Resolution is not valid");
+		return -EINVAL;
+	}
+
+	if (channels > BIT(index)) {
+		LOG_ERR("Only single channel supported");
+		return -ENOTSUP;
+	}
+
+	data->buffer = sequence->buffer;
+	data->size = sequence->buffer_size / sizeof(*data->buffer);
+
+	adc_context_lock(&data->ctx, asynchronous, sig);
+	adc_context_start_read(&data->ctx, sequence);
+	rv = adc_context_wait_for_completion(&data->ctx);
+	adc_context_release(&data->ctx, rv);
+	return rv;
+}
+
+static int adc_cc32xx_read(const struct device *dev,
+			   const struct adc_sequence *sequence)
+{
+	return cc32xx_read(dev, sequence, false, NULL);
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_cc32xx_read_async(const struct device *dev,
+				 const struct adc_sequence *sequence,
+				 struct k_poll_signal *async)
+{
+	return cc32xx_read(dev, sequence, true, async);
+}
+#endif
+
+static void adc_cc32xx_isr(const struct device *dev, int ch)
+{
+	const struct adc_cc32xx_cfg *config = dev->config;
+	struct adc_cc32xx_data *data = (struct adc_cc32xx_data *)dev->data;
+	unsigned long mask = MAP_ADCIntStatus(config->base, s_channel[ch]);
+	int cnt = 0;
+	int rv = 0;
+
+	MAP_ADCIntClear(config->base, s_channel[ch], mask);
+
+	if ((mask & ADC_FIFO_EMPTY) || !(mask & ADC_FIFO_FULL)) {
+		return;
+	}
+
+	while (MAP_ADCFIFOLvlGet(config->base, s_channel[ch])) {
+		rv += (MAP_ADCFIFORead(config->base, s_channel[ch]) >> 2) & 0x0FFF;
+		cnt++;
+	}
+
+	*data->buffer++ = rv / cnt;
+
+	if (data->buffer >= (data->repeat_buffer + data->size)) {
```

Comment:
Yeah, it was bad idea to use buffer size for sequence stop indication. Updated to rely on channel mask

> this won't work if the user provides a buffer longer than needed for the expected number of samples
Indeed it will just fill user buffer till the end and stop then.